mesonbuild/interpreter/compiler.py

```diff
@@ -797,3 +806,54 @@ def preprocess_method(self, args: T.Tuple[T.List['mesonlib.FileOrString']], kwar
         # other targets, list outputs, etc.
         private_dir = os.path.relpath(self.interpreter.backend.get_target_private_dir(tg), self.interpreter.subdir)
         return [build.CustomTargetIndex(tg, os.path.join(private_dir, o)) for o in tg.outputs]
+
+    @typed_kwargs('compiler.symbol_visibility_header',
+                  KwargInfo('api', (str, NoneType)),
+                  KwargInfo('compilation', (str, NoneType)),
+                  KwargInfo('static_compilation', (str, NoneType)),
+                  KwargInfo('static_only', bool, default=False))
+    @typed_pos_args('compiler.symbol_visibility_header', str)
+    @FeatureNew('compiler.symbol_visibility_header', '1.1.0')
+    def symbol_visibility_header_method(self, args: T.Tuple[str], kwargs: 'SymbolVisibilityHeaderKW') -> mesonlib.File:
+        if self.compiler.language not in {'c', 'cpp', 'objc', 'objcpp'}:
+            raise InterpreterException('compiler.symbol_visibility_header is supported only for C and C++ compilers')
+
+        header_name = args[0]
+
+        default_prefix = mesonlib.underscorify(self.interpreter.active_projectname.upper())
+        if default_prefix[0].isdigit():
+            default_prefix = f'_{default_prefix}'
+        api = kwargs['api'] or f'{default_prefix}_API'
+        compilation = kwargs['compilation'] or f'{default_prefix}_COMPILATION'
+        static_compilation = kwargs['static_compilation'] or f'{default_prefix}_STATIC_COMPILATION'
+        static_only = kwargs['static_only']
+
+        content = '#pragma once\n'
+        if static_only:
+            content += textwrap.dedent(f'''
+                #ifndef {static_compilation}
+                #  define {static_compilation}
+                #endif /* {static_compilation} */
+                ''')
+        content += textwrap.dedent(f'''
+            #if (defined(_WIN32) || defined(__CYGWIN__)) && !defined({static_compilation})
```

Comment:
So I would like to merge this given every project that ever exists need to do a configure_file() to get it right given the existence of static_compilation and compilation in this format string.

Reply:
Hmm I guess I am an idiot, and these would be constants and not require configure_file() :D

Reply:
Right, you can typically commit that file into your project. I personally still believe it has value in meson because it can serve as reference. You can generate it with meson initially then commit the file. It can also be useful for meson unit tests, many of them has some variant of this, it's boring to copy that file everywhere. I'm yet to see a single project who does that correctly, delegating that to meson means every dev does not have to know all the details, they can just trust meson.